src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -15,32 +15,155 @@
+   */
+  protected def execute(): Try[Unit]
 
   /**
-    * Verifies whether all preconditions are met for this specific action.
-    *
-    * @return `Success` when all preconditions are met, `Failure` otherwise
-    */
-  def checkPreconditions: Try[Unit] = Success(())
+   * Cleans up results of a previous call to run so that a new call to run will not fail because of those results.
+   *
+   * @return `Success` if the rollback was successful, `Failure` otherwise
+   */
+  protected def rollback(): Try[Unit] = Success(())
 
   /**
-    * Exectue the action.
-    *
-    * @return `Success` if the execution was successfull, `Failure` otherwise
-    */
-  def run(): Try[Unit]
+   * Run an action. First the precondition is checked. If it fails a `PreconditionsFailedException`
+   * with a report is returned. Else, if the precondition succeed, the action is executed.
+   * If this fails, the action is rolled back and a `ActionRunFailedException` with a report is returned.
+   * If the execution was successful, `Success` is returned
+   *
+   * @return `Success` if the full execution was successful, `Failure` otherwise
+   */
+  final def run: Try[Unit] = {
+    for {
+      _ <- innerCheckPreconditions.recoverWith {
+        case NonFatal(e) =>
+          Failure(PreconditionsFailedException(generateReport("Precondition failures:", e, "Due to these errors in the preconditions, nothing was done.")))
+      }
+      _ <- innerExecute().recoverWith {
+        case NonFatal(e) => List(Failure(e), innerRollback()).collectResults.recoverWith {
+          case e: CompositeException => Failure(ActionRunFailedException(generateReport("Errors in Multi-Deposit Instructions file:", e)))
+        }
+      }
+    } yield ()
+  }
+
+  private def generateReport(header: String = "", t: Throwable, footer: String = ""): String = {
+    val report = t match {
+      case ActionException(row, msg, _) => s" - row $row: $msg"
+      case CompositeException(ths) =>
+        ths.toSeq
+          .collect { case ActionException(row, msg, _) => s" - row $row: $msg" }
+          .mkString("\n")
+      case e => s" - unexpected error: ${e.getMessage}"
+    }
+
+    header.toOption.map(_ + "\n").getOrElse("") + report + footer.toOption.map("\n" + _).getOrElse("")
+  }
 
   /**
-    * Cleans up results of a previous call to run so that a new call to run will not fail because of those results.
-    *
-    * @return TODO
-    */
-  def rollback(): Try[Unit] = Success(())
+   * Compose this action with `other` into a new action such that it runs both actions in the correct order.
+   * [[checkPreconditions]] and [[execute]] are run in order, [[rollback]] is run in reversed order.
+   *
+   * @param other the second action to be run
+   * @return a composed action
+   */
+  def compose(other: Action): ComposedAction = {
+    other match {
+      case composedAction: Action#ComposedAction => ComposedAction(this :: composedAction.actions)
```

Comment:
Yup, in your project `Force` is an object, so it's an instance (singleton) as well. Here, however, `Action` is a trait rather than an object, so it's _not_ an instance. Then you use `#` instead.

See:
* http://stackoverflow.com/questions/6676048/why-does-one-select-scala-type-members-with-a-hash-instead-of-a-dot
* http://stackoverflow.com/questions/12959438/investigation-of-type-and-keywords-in-scala